index.tsx

```diff
@@ -0,0 +1,137 @@
+import {memo, useEffect, useRef, useState} from "react";
+
+enum TimeStatus {
+    Minute,
+    Second,
+    MilliSecond
+}
+
+type TimerInfo = {
+    duration:number;
+    time:TimeStatus;
+}
+
+export const CodeReview = memo(()=>{
+    const [stopWatchTime, setStopWatchTime] = useState<number>(0)
+    const [isStopWatchRunning, setIsStopWatchRunning] = useState<boolean>(false)
+    const [timerTime, setTimerTime] = useState<number>(0)
+    const [isTimerRunning, setIsTimerRunning] = useState<boolean>(false)
+    const [timerInfo, setTimerInfo] = useState<TimerInfo>({
+        duration:0,
+        time:TimeStatus.Minute
+    })
+    const stopWatchRef = useRef<ReturnType<typeof setInterval> | undefined>(undefined);
+    const timerRef = useRef<ReturnType<typeof setInterval> | undefined>(undefined);
+
+
+    useEffect(() => {
+        if(timerTime >= 0) return;
+
+        stopTimer() // 타이머 종료
+        setTimerTime(0) // 시간 초기화
+        alert('타이머가 종료되었습니다.')
+    }, [timerTime]);
+
+
+    // 스탑워치 실행 함수
+    const startStopWatch = ()=>{
+        if(isStopWatchRunning) return;
+
+        setIsStopWatchRunning(true)
+        stopWatchRef.current = setInterval(()=>{setStopWatchTime(prev=> prev + 1)},10)
+    }
+
+    // 스탑워치 정지 함수
+    const stopStopWatch = () => {
+        if(!isStopWatchRunning) return;
+
+        setIsStopWatchRunning(false)
+        clearInterval(stopWatchRef.current)
+    }
+
+    // 타이머 정지 함수
+    const stopTimer = ()=>{
+        if(!isTimerRunning) return;
+
+        setIsTimerRunning(false)
+        clearInterval(timerRef.current)
+    }
+
+    // 타이머 실행 함수
+    const startTimer = ()=>{
+        if(isTimerRunning) return
+
+        setIsTimerRunning(true)
+        const {duration, time} = timerInfo;
+
+        if(time === TimeStatus.MilliSecond) setTimerTime(duration/10)
+        if(time === TimeStatus.Second) setTimerTime(duration * 100)
+        if(time === TimeStatus.Minute) setTimerTime(duration * 6000)
+
+        timerRef.current = setInterval(()=>{
+            setTimerTime(prev => prev - 1)
+        },10)
+    }
+
+    // 시간 포맷팅 함수
+    const getTimeFormat =(duration:number):string=>{
+        const milliseconds = String(duration % 100).padStart(2, '0');
+        const seconds = String(Math.floor((duration / 100) % 60)).padStart(2, '0');
+        const minutes = String(Math.floor(duration / (100 * 60))).padStart(2, '0');
+
+        return `${minutes} : ${seconds} : ${milliseconds}`
+    }
+
+    return (
+        <>
+            <div>
+                <h1>스탑워치</h1>
+                <div>{getTimeFormat(stopWatchTime)}</div>
```

Comment:
```suggestion
    // 시간 포맷팅 함수
    const stopWatchTimeWithFormat =useMemo(() => getTimeFormat(stopWatchTime), [stopWatchTime]);

    return (
        <>
            <div>
                <h1>스탑워치</h1>
                <div>{stopWatchTimeWithFormat}</div>
```
이렇게 useMemo를 활용하는 방법도 있답니다!
다만 지금은 렌더링 성능에 큰 영향이 없을 것 같아서, 지금과 같이 표현해도 무방해보이긴 합니다 ㅎㅎ